### ISSUE TITLE
JSpecify: false positive when ? super T is on the both sides of call and T is nullable

### ISSUE THREAD

**Issue by unv-unv** [uber/NullAway#1500]
```java
@NullMarked
public class Test {
    static class Foo<T extends @Nullable Object> {
        public static <T extends @Nullable Object> Foo<T> of(Foo<? super T> foo) {
            return new Foo<>();
        }

        public Foo<T> or(Foo<? super T> other) {
            return this;
        }
    }

    // BUG: [NullAway] incompatible types: Foo<Void> cannot be converted to Foo<@Nullable Void>
    static final Foo<@Nullable Void> FOO = Foo.of(new Foo<@Nullable Void>()).or(new Foo<@Nullable Void>());
}
```

Note that if I remove call for .or() everything compiles OK.

```java
static final Foo<@Nullable Void> FOO = Foo.of(new Foo<@Nullable Void>());
```

The following one is OK too:

```java
static final Foo<@Nullable Void> FOO_BASE = Foo.of(new Foo<@Nullable Void>());
static final Foo<@Nullable Void> FOO = FOO_BASE.or(new Foo<@Nullable Void>());
```

So the problem appears only if on the both sides of .or() call is ? super T.

NullAway 0.13.1, Error Prone 2.48.0

**Comment by unv-unv** [comment]
Checked in 0.13.2. Problem is still here.

**Comment by msridhar** [comment]
Yeah, we'll update / close this issue when this one is fixed, but it will take some time.

**Comment by msridhar** [comment]
@unv-unv I wrote a Java equivalent of this test and it does not compile:

```java
class Test {

    static class Foo<T> {
    public static <T> Foo<T> of(Foo<? super T> foo) {
        return new Foo<>();
    }

    public Foo<T> or(Foo<? super T> other) {
        return this;
    }
    }
    static final Foo<String> FOO = Foo.of(new Foo<Object>()).or(new Foo<Object>());
}
```
The following modified `FOO` declaration _does_ compile:
```java
    static final Foo<String> FOO = Foo.<String>of(new Foo<Object>()).or(new Foo<Object>());
```
([Godbolt link](https://godbolt.org/z/Y875ahYMx))

Basically, `javac` cannot infer that `Foo.of(new Foo<Object>())` should have type `Foo<String>` based on `FOO`'s type, since it goes through the call to `or`.  Right now, for similar reasons, NullAway does not know that `Foo.of(new Foo<@Nullable Void>())` "should" have type `Foo<@Nullable Void>` and infers `Foo<Void>` instead.

Now, for the `Void` type, one could argue we just should always infer `@Nullable Void` and add some special treatment.  We could think about that, though it'd be low priority.  But, I don't think this issue was about the `Void` type specifically.

@unv-unv thoughts?
